tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,224 @@
+                        }
+                        DependencyGraphNode innercompNode = nodes.get(innercomponent);
+                        for (String provs : innercompType.getProvs().values()) {
+                            if (provs.equals(reqs)) {
+                                compNode.addRequires(innercompNode);
+                                innercompNode.addProvides(compNode);
+                                nodes.put(innercomponent, innercompNode);
+                                nodes.put(component, compNode);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Apply topological sort to the graph!
+        for (DependencyGraphNode m : noReqNodes) {
+            ordered.add(m.getComponentDecl());
+            for (DependencyGraphNode n : m.getProvides()) {
+                n.getProvides().remove(m);
+                if (n.getProvides().size() == 0) {
+                    ordered.add(n.getComponentDecl());
+                }
+            }
+        }
+        return ordered;
+    }
+
+    public HashMap<String, ASTComponentTypeDecl> getComponentTypes() {
+        return componentTypes;
+    }
+
+    public HashSet<String> getConnectorTypes() {
+        return connectorTypes;
+    }
+
+    public HashMap<String, ASTComponentDecl> getComponents() {
+        return components;
+    }
+
+    public HashMap<String, String> getConnectors() {
+        return connectors;
+    }
+
+    public HashMap<String, String> getEntrypoints() {
+        return entrypoints;
+    }
+
+    public HashMap<String, HashSet<String>> getAttachments() {
+        return attachments;
+    }
+
+    public HashMap<String, ASTPortDecl> getPortDecls() {
+        return portdecls;
+    }
+
+    public DeclCheckVisitor(InterpreterState state) {
+        super();
+        this.state = state;
+    }
+
+    public Object visit(ASTAttachmentDecl node, Object data) {
+        String connector = node.getConnector();
+        HashSet<String> ports = node.getAllPorts();
+        if (!connectors.containsKey(connector)) {
+            // connector not declared
+            ToolError.reportError(ErrorMessage.MEMBER_NOT_DECLARED,
+                    node.getLocation(), "connector", connector);
+        }
+        if (attachments.putIfAbsent(connector, ports) != null) {
+            // duplicate connector use
+            ToolError.reportError(ErrorMessage.DUPLICATE_CONNECTOR_USE,
+                    node.getLocation(), connector);
+        }
+
+        /* Check if ports are improperly declared */
+        for (String portFull : ports) {
+            String[] portParts = portFull.split("\\.", 2);
+            String component = portParts[0];
+            String port = portParts[1];
+            // check component declaration exists
+            // get component type
+            String type = components.get(component).getType();
```

Comment:
Same bug that I mentioned above exists here as well. Components might be declared after attachments.